# Upstream: ml-explore/mlx 017e1c3b9640483a41dcb7dd06bec60a392a42e1 (Fix slicing's non contiguous start, #4599 (CPU part; cuda/metal hunks out of scope))
# License: MIT; upstream copyright and license notices retained.
diff --git a/mlx/backend/cpu/primitives.cpp b/mlx/backend/cpu/primitives.cpp
index b4854f071..d2f5481b9 100644
--- a/mlx/backend/cpu/primitives.cpp
+++ b/mlx/backend/cpu/primitives.cpp
@@ -35,6 +35,14 @@ static std::pair<array, bool> compute_dynamic_offset(
     const Strides& strides,
     const std::vector<int>& axes,
     Stream stream) {
+  auto& encoder = cpu::get_command_encoder(stream);
+  // The kernel reads the indices contiguously
+  if (!indices.flags().row_contiguous) {
+    auto copy = contiguous_copy_cpu(indices, stream);
+    encoder.add_temporary(copy);
+    return compute_dynamic_offset(copy, strides, axes, stream);
+  }
+
   array offset({1}, int64, nullptr, {});
   bool donate = indices.is_donatable() &&
       (indices.data_size() * indices.itemsize()) >= offset.itemsize();
@@ -44,7 +52,6 @@ static std::pair<array, bool> compute_dynamic_offset(
     offset.set_data(allocator::malloc(offset.itemsize()));
   }
 
-  auto& encoder = cpu::get_command_encoder(stream);
   encoder.set_input_array(indices);
   encoder.set_output_array(offset);
   auto compute_offset =
